src/gravity-tools.c

```diff
@@ -0,0 +1,402 @@
+		fclose(fpin);
+		return EXIT_FAILURE;
+	}
+	// Get size of input file
+	fseek(fpin, 0L, SEEK_END);
+	const size_t fsize = ftell(fpin);
+	rewind(fpin);
+
+	// Compile regular expression to validate domains
+	regex_t exact_regex, abp_regex, false_positives_regex, abp_css_regex;
+	if(regcomp(&exact_regex, VALID_DOMAIN_REXEX, REG_EXTENDED) != 0)
+	{
+		printf("%s  %s Unable to compile regular expression to validate exact domains\n",
+		       over, cross);
+		fclose(fpin);
+		sqlite3_close(db);
+		return EXIT_FAILURE;
+	}
+	if(regcomp(&abp_regex, ABP_DOMAIN_REXEX, REG_EXTENDED) != 0)
+	{
+		printf("%s  %s Unable to compile regular expression to validate ABP-style domains\n",
+		       over, cross);
+		fclose(fpin);
+		sqlite3_close(db);
+		return EXIT_FAILURE;
+	}
+	if(regcomp(&false_positives_regex, FALSE_POSITIVES, REG_EXTENDED | REG_NOSUB) != 0)
+	{
+		printf("%s  %s Unable to compile regular expression to identify false positives\n",
+		       over, cross);
+		fclose(fpin);
+		sqlite3_close(db);
+		return EXIT_FAILURE;
+	}
+	if(regcomp(&abp_css_regex, ABP_CSS_SELECTORS, REG_EXTENDED) != 0)
+	{
+		printf("%s  %s Unable to compile regular expression to validate ABP-style domains\n",
+		       over, cross);
+		fclose(fpin);
+		sqlite3_close(db);
+		return EXIT_FAILURE;
+	}
+
+	// Begin transaction
+	if(sqlite3_exec(db, "BEGIN TRANSACTION;", NULL, NULL, NULL) != SQLITE_OK)
+	{
+		printf("%s  %s Unable to begin transaction to insert domains into database file %s\n",
+		       over, cross, outfile);
+		fclose(fpin);
+		sqlite3_close(db);
+		return EXIT_FAILURE;
+	}
+
+	// Prepare SQL statement
+	const char *sql = "INSERT INTO gravity (domain, adlist_id) VALUES (?, ?);";
+	if(sqlite3_prepare_v2(db, sql, -1, &stmt, NULL) != SQLITE_OK)
+	{
+		printf("%s  %s Unable to prepare SQL statement to insert domains into database file %s\n",
+		       over, cross, outfile);
+		fclose(fpin);
+		sqlite3_close(db);
+		return EXIT_FAILURE;
+	}
+
+	// Bind adlistID
+	const int adlistID = atoi(adlistIDstr);
+	if(sqlite3_bind_int(stmt, 2, adlistID) != SQLITE_OK)
+	{
+		printf("%s  %s Unable to bind adlistID to SQL statement to insert domains into database file %s\n",
+		       over, cross, outfile);
+		fclose(fpin);
+		sqlite3_close(db);
+		return EXIT_FAILURE;
+	}
+
+	// Parse list file line by line
+	char *line = NULL;
+	size_t lineno = 0;
+	size_t len = 0;
+	ssize_t read = 0;
+	size_t total_read = 0;
+	int last_progress = 0;
+	char *invalid_domains_list[MAX_INVALID_DOMAINS] = { NULL };
+	unsigned int invalid_domains_list_len = 0;
+	unsigned int exact_domains = 0, abp_domains = 0, invalid_domains = 0;
+	while((read = getline(&line, &len, fpin)) != -1)
+	{
+		// Update total read bytes
+		total_read += read;
+		lineno++;
+
+		// Remove leading whitespace
+		// isspace() matches spaces, tabs, form feeds, line breaks,
+		// carriage returns, and vertical tabs
+		while(isspace(line[0]))
+			line++;
+
+		// Skip comments
```

Comment:
Here is some duplicate code L166-186. This is still part of `gravity.sh` 
`gravity_ParseFileIntoDomains()`

https://github.com/pi-hole/pi-hole/blob/73733308ba0f1c4e2cc5fb0c4aac6a9027349e48/gravity.sh#L639-L664

Does not hurt to keep it, just wanted to report.

Reply:
If it is duplicated, we should remove it from FTL for minimal speed gain.

Reply:
Actually, rethinking this, we can also remove `gravity_ParseFileIntoDomains()` from `gravity.sh` and use the code here that should be a lot faster.

Reply:
Then this would be something for https://github.com/pi-hole/pi-hole/pull/5275